src/dhtproto/client/request/internal/GetChannels.d

```diff
@@ -0,0 +1,367 @@
+        Returns:
+            true to continue handling the request (OK status); false to abort
+            (error status)
+
+    ***************************************************************************/
+
+    private bool handleStatusCode ( ubyte status )
+    {
+        auto getchannels_status = cast(RequestStatusCode)status;
+
+        if ( GetChannels.handleGlobalStatusCodes(getchannels_status,
+            this.context, this.conn.remote_address) )
+        {
+            return false; // Global code, e.g. request/version not supported
+        }
+
+        // GetChannels-specific codes
+        with ( RequestStatusCode ) switch ( getchannels_status )
+        {
+            case Started:
+                // Expected "request started" code
+                return true;
+
+            case Error:
+                // The node returned an error code. Notify the user and
+                // end the request.
+                GetChannels.Notification n;
+                n.node_error = RequestNodeInfo(
+                    this.context.request_id, conn.remote_address);
+                GetChannels.notify(this.context.user_params, n);
+                return false;
+
+            default:
+                // Treat unknown codes as internal errors.
+                goto case Error;
+        }
+
+        assert(false);
+    }
+
+    /***************************************************************************
+
+        Handler policy, called from AllNodesRequest template to run the
+        request's main handling logic.
+
+    ***************************************************************************/
+
+    private void handle ( )
+    {
+        // Receive the incoming record stream.
+        bool finished;
+        do
+        {
+            this.conn.receive(
+                ( in void[] payload )
+                {
+                    finished = this.handleMessage(payload);
+                }
+            );
+        }
+        while ( !finished );
+    }
+
+    /***************************************************************************
+
+        Handles a single message received from the node.
+
+        Params:
+            payload = message payload
+
+        Returns:
+            true if the message indicates that the request is finished
+
+    ***************************************************************************/
+
+    private bool handleMessage ( Const!(void)[] payload )
+    {
+        bool finished;
+
+        auto msg_type =
+            *this.conn.message_parser.getValue!(MessageType)(payload);
+        with ( MessageType ) switch ( msg_type )
+        {
+            case ChannelName:
+                Const!(void)[] channel;
+                this.conn.message_parser.parseBody(payload, channel);
+                Notification notification;
+                notification.received =
+                    RequestDataInfo(context.request_id, channel);
+                GetChannels.notify(this.context.user_params, notification);
```

Comment:
Hm. Does this imply user receiving duplicate channel names in the notifier?

Reply:
Yes. This request has always worked like that.

Maybe we should think of something better, though. Hadn't occurred to me until you mentioned this :(

Reply:
Yeah, I think it shouldn't be hard to have a set of channels that user already received notification for inside `SharedResources`.

Reply:
Will do. This will require two things:
1. A template in swarm for structs wrapping a `void[]` and treating it as an array of another type.
2. Acquiring two arrays. Append channel names to one and store slices to the first in the second.

Reply:
This is problematic at the moment. See https://github.com/sociomantic-tsunami/dhtproto/issues/34.

Suggest continuing to review / merge this PR as-is.